Add ioctl calculation, add tests

XANi opened the pull request:
Hi,

so I did some testing where it comes to ioctl numbers (which are currently hardcoded) and they seem to be the same on at least arm64 and amd64. I've added some helper code to generate them from scratch so at the very least there is parity with the kernel code

While testing those functions I've noticed that the Sizeof() golang data structures is different than the C equivalents and i've tracked it to `int` being 32 bit on 64 bit platforms in C so I changed it around a bit.

I've also added a bunch of tests both for the ioctl functions and for the GPIO operations. At the very least they pass fine on rPi3:

```
root@hq-upsmgr:~/src/gpio-cdev-go# export GPIO_DEV_PATH="/dev/gpiochip0" ;export GPIO_TEST_PIN="19";export GPIO_TEST_PIN_LOOP="16"
root@hq-upsmgr:~/src/gpio-cdev-go# go test -v
=== RUN   TestGPIO
dev: /dev/gpiochip0 in: 19, out: 16
read: 0
read: 0
read: 1
read: 0
--- PASS: TestGPIO (0.00s)
=== RUN   TestGPIOEvent
setting event trigger in 10ms
setting wait for 10s
triggered after 10.473636ms
event: {Timestamp:1574246665685767805 ID:1 _pad:0}
--- PASS: TestGPIOEvent (0.01s)
=== RUN   TestIoctlHelper
--- PASS: TestIoctlHelper (0.00s)
PASS
ok  	github.com/temoto/gpio-cdev-go	0.055s
```

XANi committed:
Add ioctl calculation, add tests

Changes:

* Fd changed to int32 - linux uses int-sized Fd, which on 64 bit platforms *in C* is int32, while *in Go* it is int64 - this is required to keep size/alignment of Go structs be in par with the C ones
* Added functions to calculate ioctl numbers same way as kernel does - so far I've checked amd64 and arm64 and it looks the same as previous values
* Dropped Op/uint for ioctl values, it gets converted to uintptr anyway.
* Tests for ioctl calculation, GPIO and evented GPIO

temoto commented on readme.md:
```diff
@@ -15,6 +15,17 @@ Ultimate success would be to merge this functionality into periph.io lib.
 - may leak `req.fd` descriptors, TODO test
 - 64-bit system likely have different ioctl numbers, please try and write me back
 
+# Testing:
```
Please make it just `# Testing` because punctuation (`:` `.` etc) is not used in headers.

temoto commented on gpio_test.go:
```diff
@@ -0,0 +1,158 @@
+	"github.com/stretchr/testify/assert"
+)
+
+// valudate and extract test setup
```
typo: validate

temoto commented on gpio_test.go:
```diff
@@ -0,0 +1,158 @@
+		return "",0, 0
+	}
+	testPinIn, err := strconv.Atoi(os.Getenv("GPIO_TEST_PIN"))
+	if (testPinIn < 1 || err != nil) {
```
Pretty sure it's not the scope of current library to validate pin numbers people actually may have.

XANi replied:
For some reason I thought that GPIO pins start at 1, I'll go and fix that assumption

temoto commented on gpio_test.go:
```diff
@@ -0,0 +1,158 @@
+func TestGPIO(t *testing.T) {
+	a := assert.New(t)
+	testDevice, testPin, testPinLoop := getGpioSetup(t)
+	fmt.Printf("dev: %s in: %d, out: %d\n", testDevice,testPin,testPinLoop)
```
`t.Logf`

temoto commented on gpio_test.go:
```diff
@@ -0,0 +1,158 @@
+	// no test pin provided, skip
+	if testPin == 0 { return }
+	chiper, err := Open(testDevice, "go-test")
+	if !a.Nil(err,"Open should succeed") { return }
```
Please make `assert.Nil(t, err)` or at least `assert:=assert.New(t)`
current `a.Nil` is not immediately easy to understand.

temoto commented on gpio_test.go:
```diff
@@ -0,0 +1,158 @@
+	a.Nil(err,"Read should succeed")
+	fmt.Printf("read: %v\n",data.Values[0])
+
+	// no loopback pin defined ,exit
```
`defined, exit`

but whole comment line is redundant here

temoto commented on gpio_test.go:
```diff
@@ -0,0 +1,158 @@
+
+// valudate and extract test setup
+func getGpioSetup(t *testing.T) (gpiodev string, gpioin uint32, gpioout uint32) {
+	if len(os.Getenv("GPIO_DEV_PATH")) == 0 {
```
```go
func getEnv(t *testing.T) (dev string, pinIn uint32, pinOut uint32) {
	dev = os.Getenv("GPIO_TEST_DEV")
	inStr := os.Getenv("GPIO_TEST_PIN_INPUT")
	outStr := os.Getenv("GPIO_TEST_PIN_OUTPUT")
	if dev == "" {
		t.SkipNow(...)
	}
```

- getGpioSetup -> getEnv, because everything here is about gpio already, this word adds noise to reading
- return variables declared but not used
- read environment once
- consistent GPIO_TEST_ prefix
- GPIO_DEV_PATH -> GPIO_TEST_DEV because PATH feels redundant here, up to you
- important to mark GPIO_TEST_PIN_INPUT/OUTPUT as they are used
- "idiomatic" to check `str == ""` without `len()`
- SkipNow, Fatalf to save on checking in tests below

XANi replied:
> return variables declared but not used

It is for IDE documentation purpose. Most IDEs will show the function prototype as hint

temoto replied:
Yeah I agree it's good documentation, especially when types are not so obvious like in `Atoi() string,error`
But since you declared return variables, might as well use them.
Updated example above.

XANi replied:
> important to mark GPIO_TEST_PIN_INPUT/OUTPUT as they are used

My reasoning behind the name is that it isn't "just" output but specifically "output looped to the other pin, so it is is immediately apparent from the variable name that you need to do more than just find first free GPIO pin.

Also there would be a conflict if say there would be tests added that use the dummy `gpio-mockup` driver as there is no way to loop pins to eachother in it (afaik)

Not that I *particularly* care about the environmental names.

I've changed variable names in code so they are indicative of the purpose the pin is used for at the given moment.

temoto commented on gpio_test.go:
```diff
@@ -0,0 +1,158 @@
+package gpio
+
+import (
+	"fmt"
```
Definitely `fmt.Print -> t.Log`

temoto commented on ioctl.go:
```diff
@@ -0,0 +1,60 @@
+package gpio
```
Please make everything in this file not exported, like `func Io -> io` because it's not part of library API.

temoto commented on ioctl_linux_test.go:
```diff
@@ -0,0 +1,57 @@
+
+
+func TestIoctlHelper(t *testing.T) {
+	a := assert.New(t)
```
Please, full `assert.Equal(t, ...)` or `assert := assert.New()`

temoto reviewed (changes requested):
Absolutely fantastic, thank you so much.

Please run `go fmt` and minor style changes or I can do that for you if you don't mind commit changed.

temoto commented:
> I've added some helper code to generate them from scratch so at the very least there is parity with the kernel code

It's also present in repo as `print-ioctl.c.go-skip` but maybe I should've mention it in readme or somewhere.

temoto commented on ioctl_linux_test.go:
```diff
@@ -0,0 +1,57 @@
+
+func TestIoctlHelper(t *testing.T) {
+	a := assert.New(t)
+	// #define GPIO_GET_CHIPINFO_IOCTL _IOR(0xB4, 0x01, struct gpiochip_info)
```
In this situation, table tests work best:

```go
func TestIoctlHelper(t *testing.T) {
	cases:= []struct {
		name string
		expect uintptr
		actual uintptr
	}{
		{"GPIO_GET_CHIPINFO_IOCTL", 0x8044b401, IoR(0xb4, 0x01, unsafe.Sizeof(ChipInfo{}))},
		// ...
	}
	for _, c := range cases {
		assert.Equal(t, c.expect, c.actual, c.name)
	}
}
```

XANi committed:
formatting and naming fixes

XANi commented:
I've pushed changes, I think that should be most of the issues fixed

temoto commented on gpio_test.go:
```diff
@@ -0,0 +1,173 @@
+)
+
+// validate and extract test setup
+func getEnv(t *testing.T) (gpioDev string, gpioIn uint32, gpioOut uint32, hasGpio bool, hasLoopback bool) {
```
Now that there are so many return values, maybe struct is better. Your choice.

Examples:
- https://github.com/temoto/vender/blob/062e69b44c6a9bd1eeb613a86ae75227cefd5887/head/tele/tele_test.go
- https://github.com/temoto/vender/blob/0a1110c8ce7d0f713d29dfbda1a75b01f0fbf9ba/head/ui/util_test.go#L18

temoto commented on gpio_test.go:
```diff
@@ -0,0 +1,173 @@
+	inStr := os.Getenv("GPIO_TEST_PIN")
+	outStr := os.Getenv("GPIO_TEST_PIN_LOOP")
+	if gpioDev == "" {
+		t.Skip("Please set GPIO_TEST_DEV, GPIO_TEST_PIN to run tests ")
```
t.SkipNow, t.Fatalf could eliminate `if hasGpio` checks below.

temoto commented on gpio_test.go:
```diff
@@ -0,0 +1,173 @@
+		return
+	}
+	chiper, err := Open(testDevice, "go-test")
+	if !assert.Nil(err, "Open should succeed") {
```
There is `require.Nil,Equal,...` set of functions that act as SkipNow, Fatal to avoid 3 lines on if statement.

temoto replied:
I mean `require.NoError(err, "Open should succeed")` is equivalent to

```
if !assert.NoError(err, "Open should succeed") {
	t.FailNow()
}
```

But don't worry if you don't want. Current version is great.

XANi replied:
If I knew it existed I would've used it :) I've used goconvey before (mostly because of UI) but it kinda...fell apart when touching modules so I do not have much experience with other frameworks yet.

temoto commented on ffi_ioctl_32.go:
```diff
@@ -1,13 +1,13 @@
-	GPIO_GET_LINEEVENT_IOCTL         Op = 0xc030b404
-	GPIOHANDLE_GET_LINE_VALUES_IOCTL Op = 0xc040b408
-	GPIOHANDLE_SET_LINE_VALUES_IOCTL Op = 0xc040b409
+// From <include/uapi/linux/gpio.h
```
I'm afraid of possible problems with these values changed. Now that you added test that verifies them against ioR, maybe it's fine to revert to const?

XANi replied:
Either way it will get caught in tests so up to you

temoto replied:
I mean with `var GPIO_GET_LINEEVENT_IOCTL` it's possible to write external code like this: `gpio.GPIO_GET_LINEEVENT_IOCTL = 1234`

Please change them to const.

temoto commented:
Also golangci-lint is not happy about few things. https://travis-ci.org/temoto/gpio-cdev-go/jobs/614524832

XANi committed:
Test refactor

* table test for IOCTLs
* hardcode constants but check them in tests
* test err before deferring Close()
* split gpio config into function to extract them and messages about missing env variables in separate text
* fail early if open failed

XANi commented:
I've refactored ioctl tests to use array and test the constants, and fixed a bunch of other stuff.

I've also split the "get test config from env part" and "tell user what they need to setup into separate function/test so there is no duplicat

> Also golangci-lint is not happy about few things. https://travis-ci.org/temoto/gpio-cdev-go/jobs/614524832

I didn't cull unused functions/constants from ioctl.go as they might be used in the future if the kernel api expands, and it wasn't my code in the first place so didn't wanted to touch it if not necessary

temoto reviewed (approved):
Thank you very much.

temoto reviewed (changes requested):
CI fails with

```
$HOME/bin/golangci-lint run

ioctl.go:12:2: const `typeMask` is unused (unused)

	typeMask      = (1 << typeBits) - 1

	^

ioctl.go:32:6: func `ioNone` is unused (unused)

func ioNone(t, nr uintptr) uintptr {

     ^

ioctl.go:42:6: func `ioW` is unused (unused)

func ioW(t, nr, size uintptr) uintptr {

     ^

ioctl.go:15:2: const `directionMask` is unused (unused)

	directionMask = (1 << directionBits) - 1

	^

ioctl.go:13:2: const `numberMask` is unused (unused)

	numberMask    = (1 << numberBits) - 1

	^

ioctl.go:14:2: const `sizeMask` is unused (unused)

	sizeMask      = (1 << sizeBits) - 1

	^
```

temoto commented:
Sorry for pedantic approach, this comment will make linter calm.

```
var foo string //lint:ignore U1000 unused
```

XANi committed:
cull redundant if statments in tests, drop unused consts

XANi commented:
Replaced extra ifs with requires

> CI fails with

I know, I left that on purpose as that was a code from the external repo I've pulled that might get useful if kernel api ever gets extended. Up to you whether it should be there or be removed/commented out

temoto committed:
lint unused

temoto committed:
to be squashed

temoto committed:
lint unused

temoto committed:
lint unused

temoto commented:
@XANi are you ok with me adding commits?

XANi commented:
Sure, It's your repo ;) But probably needs a readme change too, at the very least it was tested on arm64 (test hardware was rPi 3 B+, i was writing a driver for [UPS Pico](https://thepihut.com/products/ups-pico?variant=20063191203902)

temoto committed:
squash

temoto commented:
@XANi you mean remove 64bit ioctl from possible issues?

XANi commented:
@temoto yes, that's why I renamed the file defining them, they work fine on arm64 (and the C file you've provided shows that amd64 has the same too)

temoto commented:
Okay it's complete then. Again, thank you very much, great job.
